File: 2019/day6/day6.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

// type Planet interface {
// 	distance()
// 	indirectOrbits()
// 	directOrbit()
// }

type planet struct {
	name     string
	orbits   *planet
	children []*planet
}

func (p *planet) distance(name string) int {
	planets := p.indirectOrbits()
	for i, planet := range planets {
		if strings.EqualFold(planet.name, name) {
			return i
		}
		i++
	}
	return -1
}

func (p *planet) indirectOrbits() []*planet {
	iPlanets := make([]*planet, 0, 2)
	planet := p.orbits
	for {
		if planet == nil {
			break
		}
		iPlanets = append(iPlanets, planet)
		planet = planet.orbits
	}
	return iPlanets
}

func (p *planet) directOrbit() *planet {
	return p.orbits
}

func new(name string) *planet {
	return &planet{
		name: name,
	}
}

func setup() (map[string]*planet, error) {
	data, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	planets := make(map[string]*planet)

	for _, orbit := range strings.Split(string(data), "\n") {
		names := strings.Split(orbit, ")")
		for _, name := range names {
			if _, ok := planets[name]; !ok {
				p := new(name)
				planets[name] = p
			}
		}
		p1 := planets[names[0]]
		p2 := planets[names[1]]
		p2.orbits = p1
	}

	return planets, nil
}

func findIntersection(p1, p2 *planet) *planet {
	indirects := p2.indirectOrbits()
	for _, a := range p1.indirectOrbits() {
		for _, b := range indirects {
			if a.name == b.name {
				return a
			}
		}
	}
	return nil
}

func main() {
	p, _ := setup()
	total := 0
	for _, planet := range p {
		total += len(planet.indirectOrbits())
	}
	fmt.Println(total)
	y := p["YOU"]
	s := p["SAN"]
	i := findIntersection(y, s)
	fmt.Println(y.distance(i.name) + s.distance(i.name))
}

File: 2019/day6/day6.py
#!/usr/bin/env python3


class Planet:

    def __init__(self, name):
        self._orbits = None
        self.children = []
        self.name = name

    @property
    def orbits(self):
        return self._orbits

    @orbits.setter
    def orbits(self, x):
        if self._orbits is not None:
            print('Overwriting {} orbit with {}'.format(self, x))
        self._orbits = x
        x.add_child(self)

    def add_child(self, x):
        self.children.append(x)

    def __len__(self):
        if self._orbits:
            return 1
        return 0

    def indirect_orbits(self):
        planet = self.orbits
        while True:
            if planet is None:
                break
            yield planet
            planet = planet.orbits

    def distance_to(self, planet_name):
        total = 0
        if self.name == planet_name:
            return total
        for planet in self.indirect_orbits():
            if planet.name == planet_name:
                return total
            total += 1
        return -1

    def __str__(self):
        return self.name

    def __repr__(self):
        return 'Planet({})'.format(self.name)

    def __hash__(self):
        return hash(self.name)

    def __eq__(self, other):
        return self.name == other.name and len(self) == len(other)


def find_intersection(planetA, planetB):
    b = list(planetB.indirect_orbits())
    for a in planetA.indirect_orbits():
        if a in b:
            return a


if __name__ == '__main__':
    with open('input.txt') as f:
        orbits = f.read().split()
    planets = {}
    for orbit in orbits:
        planetA, planetB = orbit.split(')')
        for planet in [planetA, planetB]:
            try:
                planets[planet]
            except KeyError:
                p = Planet(planet)
                planets[planet] = p
        b = planets[planetB]
        a = planets[planetA]
        b.orbits = a
    you = planets['YOU']
    san = planets['SAN']
    intersection = find_intersection(you, san)

    print(you.distance_to(intersection.name) +
          san.distance_to(intersection.name))
